### ISSUE TITLE
Document the fixme/todo comment standard and natural indentation for narrative docs

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7786]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)| @adiroiban reported|
|-|-|
|Trac ID|trac#7786|
|Type|enhancement|
|Created|2015-02-07 15:56:18Z|
|Branch|https://github.com/twisted/twisted/tree/fixme-standard-7786|

current documentation does not dictate how to reference tickets and does not provide an example of such a comment.

I think that we should have a standard format.

This could be use later to automate various checks... check that we are not adding references for closed tickets... report todo tickets

------

Also, current standard does not talk about semantic line breaks in narrative docs.

<details><summary>Searchable metadata</summary>

```
trac-id__7786 7786
type__enhancement enhancement
reporter__adiroiban adiroiban
priority__normal normal
milestone__ 
branch__branches_fixme_standard_7786 branches/fixme-standard-7786
branch_author__adiroiban adiroiban
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1423324578945025 1423324578945025
changetime__1426978923026519 1426978923026519
version__None None
owner__adiroiban adiroiban

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_1"></a>|@adiroiban commented|
|-|-|

(In [43878]) Branching to fixme-standard-7786.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_2"></a>|@adiroiban removed owner|
|-|-|

I tried to update the docs for comments. PEP8 pretty much covers everything.

I put an example for a fixme reference.

How to check
```
cd docs
make html
# Check that sphinx does not complain about anything then open in browser
open _build/html/core/development/policy/coding-standard.html#comments
```

Please take a look and let me know if there is something that needs to be changes.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph set owner to @adiroiban|
|-|-|

Looks like a good change; please land it.

The "a different docstring standard" mention should probably be a link to the relevant Twisted standard document though.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

The docs don't seem to follow the semantic newlines guidelines (it is not ironic that an update to the coding standard document does not follow the coding standard: it is just a sad commentary on the efficacy of documenting the coding standard as we do).

Please fix the formatting before merging.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_5"></a>|@adiroiban removed owner|
|-|-|

Thanks for the review.

Where are the semantic newlines guidelines documented / located ?

I have updated my changes. 

As I have not found guidance in this doc I have also updated the old "Whitespace" section into a "Indentation and line breaks".

I think that semantic newlines guidelines should be located in the same doc and not somewhere else.

The later changes are in a separate commit. For now I have extended the scope of this ticket, but if required I can also open a new ticket.

Please let me know if my understanding and example of semantic line breaks is valid.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

> Where are the semantic newlines guidelines documented / located ?

https://twistedmatrix.com/documents/current/core/development/policy/doc-standard.html#xhtml-layout

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_7"></a>|@adiroiban commented|
|-|-|

Thanks. I just read "XHTML Layout" and scanned the content of the other sections and completely ignored the whole document as I think XHTML is no longer used.

Do we need to keep a separate document for doc-standard? Maybe it can be merged into the coding standard.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph commented|
|-|-|

Replying to [adiroiban](#note_7):
> Thanks. I just read "XHTML Layout" and scanned the content of the other sections and completely ignored the whole document as I think XHTML is no longer used.

This should definitely  be fixed, and this document shouldn't talk about XHTML.

> Do we need to keep a separate document for doc-standard? Maybe it can be merged into the coding standard.

The standards for narrative docs and code are radically different (for example, I would not want someone to be confused and think that semantic newlines are the standard in epytext, it's bad enough that we have them for ReST ;-)) so I think separate documents make sense.  I would not be opposed to a change that just made them separate sections of the same document though, if you strongly feel that would make them easier to read.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_9"></a>|@adiroiban commented|
|-|-|

It depends to what question you are looking to answer:

* What are the newlines rules for Twisted project?
* What are the newlines rules for Twisted python code?
* What are the newlines rules for Twisted test code?
* What are the newlines rules for Twisted narrative .RST docs?
* What are the newlines rules for Twisted news fragments?
* What are the newlines rules for Twisted Y component or X file format or Z language?

Most of the time I look for an answer for "What are the newlines rules for Twisted project?".

In the answer of this question I like to see all rules and all exceptions so that I don't have to read N other files and see how to handle new lines.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph commented|
|-|-|

Replying to [adiroiban](#note_9):

> In the answer of this question I like to see all rules and all exceptions so that I don't have to read N other files and see how to handle new lines.

For what it's worth, I think semantic newlines are dumb and I think we should just wrap .rst documents to 79 columns like everyone else, but I was overruled when this discussion was taking place :).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_11"></a>|@glyph commented|
|-|-|

I wrote a thing to the mailing list because I didn't even know how to write a review for this.

https://twistedmatrix.com/pipermail/twisted-python/2015-March/029224.html

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_12"></a>|@adiroiban commented|
|-|-|

Beside

```
Documentation should be formatted with a single sentence or clause per line. This results in diffs that are easier to read, making documentation maintenance easier.
```

https://twistedmatrix.com/documents/current/core/development/policy/doc-standard.html

only talks about lore, so I guess that is safe to just remove it and include the semantic newlines in the main code standard.

----

From my understanding semantic newlines are the current standard for documentation.
This review should be done with that in mind.

Later when we decide to drop that standard, we can create a new ticket.

I don't think that this should stop us from reviewing this ticket and fixing the current documentation.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_13"></a>|@glyph commented|
|-|-|

Replying to [adiroiban](#note_12):

> I don't think that this should stop us from reviewing this ticket and fixing the current documentation.

In principle I agree, but in practice I absolutely hate the example that is presented in your documentation, and I don't want people to copy it, so I'd prefer to reject it.  Yet, it is an accurate reflection of the current standard, so I have no basis to do so.  So I didn't see a path forward other than to change the standard :).

Another reviewer should feel free to accept this if they feel it passes muster.  I just couldn't see past the fact that the example is duplicative with the invalid "XHTML" stuff, creating more places to update the documentation when we change this, and also looks terrible.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_14"></a>|@glyph commented|
|-|-|

(Also, I think "Modern code-oriented text editor are very good at wrapping and arranging long lines." is a factually incorrect statement.  Looking at this in a half a dozen text editors, they do a half-baked job at best, and zero of them can even be configured to display "semantic newlines" as a contiguous paragraph.)

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_15"></a>|@adiroiban commented|
|-|-|

Thanks for the review.


I have removed the semantic newline section from doc-standard and reverted all the newline changes.

Moved changes talking about newline in documentation inside the writing-standard and removed the doc-standard.

The semantic newline documentation should no longer be duplicate.

Please take another look and let me know what needs to be change in order to have the documentation fixed.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_16"></a>|@glyph set owner to @adiroiban|
|-|-|

Thanks for moving forward on this, adi.  Sorry for my not entirely helpful comments earlier.

Here's a real review.

  1. "It does not has to be" is a subject/verb plurality disagreement; it should be "It does not have to be".
  2. Similarly, "Modern code-oriented text editor are" should be "Modern code-oriented text editor**s** are".
  3.
    1. For describing semantic newlines, I think we should start off with a description of what it ''is'' rather than what it's called. "semantic newlines" is very jargon-y; there are only a few references on the web, mainly [https://github.com/brandon-rhodes/sphinx-tutorial/blob/master/handout/using-sphinx.rst#hints this one].  Instead, just start out with "Make lines short, and break lines at natural places" and then mention afterwards that we call this "semantic newlines".
    2. Similarly, "narrative documentation" may not be entirely obvious from context to a first-time reader.  Be sure to note that this applies to reStructuredText documents for sphinx but ''not'' for docstrings.

If you address these minor changes and get a clean documentation build I think this can land.

Thanks again.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_17"></a>|@adiroiban removed owner|
|-|-|

Thanks for the review. Happy to see some progress on this :)

Please check latest version. Branch sent to buildbot.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_18"></a>|@hawkowl set owner to @adiroiban|
|-|-|

Thanks for this, Adi.

- docs/core/development/policy/coding-standard.rst:590 -- Could you apply the semantic newlines stuff to these lines?
- docs/core/development/policy/coding-standard.rst:595 -- "A brief text" -> "A brief amount of text"

Otherwise, it is a vast improvement. Please fix these two small issues and land.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_19"></a>|@adiroiban set status to `closed`|
|-|-|

(In [44179]) Merging fixme-standard-7786: Document the fixme/todo comment standard and natural indentation for narrative docs.

Author: adiroiban
Reviewers: glyph, hawkowl
Fixes: [#7786](https://github.com/twisted/twisted/issues/7786)
